File: src/main/scala/examples/crv15/acsvreader/scala/Main.scala
package examples.crv15.acsvreader.scala

// -----------------
// ---  LogFire: ---
// -----------------

class Monitor {
  val PRINT = true

  val nameOfPosition: Map[Int, Symbol] =
    Map(1 -> 'one, 2 -> 'two, 3 -> 'three, 4 -> 'four, 5 -> 'five,
      6 -> 'six, 7 -> 'seven, 8 -> 'eight, 9 -> 'nine)

  def addMapEvent(map: Map[Symbol, Any]) {
    if (PRINT) {
      for ((key, value) <- map) print(s"$key = $value ")
      println()
    }
  }

  def terminate() {
    println("END")
  }
}


// ---------------------
// --- The property: ---
// ---------------------

class M extends Monitor 


// ------------------------
// --- Running LogFire: ---
// ------------------------

object Main {
  import org.apache.commons.csv._

  implicit def extendMonitor(monitor: Monitor) = new {
    def submit(record: CSVRecord) {
      var map: Map[Symbol, Any] = Map();
      map += 'kind -> record.get(0)
      for (i <- 1 until record.size()) {
        map += monitor.nameOfPosition(i) -> record.get(i)
      }
      monitor.addMapEvent(map);
    }
  }

  def main(args: Array[String]) {
    import java.io._
    import scala.collection.JavaConverters._

    val monitor = new M
    val in: Reader = new BufferedReader(new FileReader(args(1)))
    val records: Iterable[CSVRecord] = CSVFormat.DEFAULT.withHeader().parse(in).asScala
    records foreach monitor.submit
    monitor.terminate()
    in.close()
  }
}

File: src/main/scala/examples/crv15/acsvreader/java/Main.java
package examples.crv15.acsvreader.java;

import java.io.*;
import java.util.*;
import org.apache.commons.csv.*;

class Monitor {
	String monitorName;

	public static Boolean PRINT = false;

	public Monitor(String monitorName) {
		this.monitorName = monitorName;
		System.out.println("BEGIN");
	}

	void addMapEvent(Map<String, String> map) {
		if (PRINT) {
			for (String key : map.keySet()) {
				System.out.print(key + "=" + map.get(key) + " ");
			}
			System.out.println();
		}
	}
	
	void terminate() {
		System.out.println("END");
	}
}

class CSVMonitor extends Monitor {
	public CSVMonitor(String monitorName) {
		super(monitorName);
	}

	private String symbolOf(int index) {
		String result = "missed";
		switch (index) {
		case 1:
			result = "one";
			break;
		case 2:
			result = "two";
			break;
		case 3:
			result = "three";
			break;
		case 4:
			result = "four";
			break;
		case 5:
			result = "five";
			break;
		case 6:
			result = "six";
			break;
		case 7:
			result = "seven";
			break;
		case 8:
			result = "eight";
			break;
		case 9:
			result = "nine";
			break;
		}
		return result;
	}

	public void submit(CSVRecord record) {
		Map<String, String> map = new HashMap();
		map.put("kind", record.get(0));
		for (int i = 1; i < record.size(); i++) {
			map.put(symbolOf(i), record.get(i));
		}
		addMapEvent(map);
	}
}

public class Main {
	public static void main(String[] args) throws IOException {
		CSVMonitor monitor = new CSVMonitor(args[0]);
		Reader in = new BufferedReader(new FileReader(args[1]));
		Iterable<CSVRecord> records = CSVFormat.DEFAULT.withHeader().parse(in);
		for (CSVRecord record : records) {
			monitor.submit(record);
		}
		monitor.terminate();
		in.close();
	}
}
